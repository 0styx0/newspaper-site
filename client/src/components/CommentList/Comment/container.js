import React from 'react';
import {jwt} from '../../jwt';
import fetchFromApi from '../../../helpers/fetchFromApi';

import PropTypes from 'prop-types';

import Comment from './';


/**
 * @param props - @see CommentContainer.propTypes
 */
class CommentContainer extends React.Component {

    constructor(props) {
        super(props);

        this.delete = this.delete.bind(this);

        this.state = {
            content: this.props.content,
            author: this.props.author
        }
    }

    /**
     * Sends request to server to delete the comment
     * 
     * Rerenders the Comment making it clear that it was deleted
     */
    delete() {

        this.setState({
            content: 'deleted',
            author: 'Deleted User'
        });

        const info = {
            id: this.props.id
        }

        fetchFromApi("comment", "delete", info);
    }

    render() {

        const commentProps = {
            author: this.state.author,
            profileLink: this.props.profileLink,
            content: this.state.content
        }

        // if already deleted, or user not authorized to delete,
        // or it was just created so info needed to delete isn't there yet
        if (this.props.content !== "deleted" &&
            (jwt.level > 2 || jwt.id === this.props.authorid) &&
            this.props.id
            ) {

            commentProps.deleteButton = this.delete;
        }

        return <Comment
                 {...commentProps}
               />
    }
}

// @see Comment 's propTypes, which are much the same
CommentContainer.propTypes = {

    profileLink: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    authorid: PropTypes.number.isRequired,
    id: PropTypes.number
}

export default CommentContainer;