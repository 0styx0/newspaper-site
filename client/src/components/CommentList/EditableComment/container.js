import React from 'react';
import PropTypes from 'prop-types';
import fetchFromApi from '../../../helpers/fetchFromApi';

import EditableComment from './';

export default class EditableCommentContainer extends React.Component {

    constructor() {
        super();

        this.save = this.save.bind(this);

        this.state = {
            content: ''
        }
    }

    /**
     * Sends comment to be saved to database
     */
    save() {

        this.props.addToList(this.state.content);

        const info = {
            issue: this.props.issue,
            url: this.props.name,
            content: this.state.content
        }

        fetchFromApi("comment", "post", info);
    }

    render() {

        return <EditableComment
                 onSubmit={this.save}
                 onBlur={e => this.setState({content: e.target.innerHTML})}
               />
    }

}


EditableCommentContainer.propTypes = {

    issue: PropTypes.string.isRequired, // issue of parent article
    name: PropTypes.string.isRequired, // name (url) of article (ex: name in http://localhost:3001/issue/1/story/name)
    addToList: PropTypes.func.isRequired // callback where content is passed into after user submits the comment
}