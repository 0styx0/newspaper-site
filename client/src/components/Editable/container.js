import React from 'react';
import PropTypes from 'prop-types'

import Editable from './';

/**
 * @prop
 * @prop canEdit - boolean
 * @prop buttons - if "all" (default), show all buttons. If "basic" show subset
 * @prop submit - runs when the submit button is clicked. If this is not given, nothing will happen when submit is clicked
 *
 * @return lets content be edited and renders a bar of buttons that can edit the html of props.children if props.canEdit = true
 */
class EditableContainer extends React.Component {

    constructor() {
        super();

        this.handleEdits = this.handleEdits.bind(this);
    }

    handleEdits(event) {

        if (event.target.classList.contains('hideFromPreview')) {
            return this.toggleClassOnElementSelected("previewHidden");
        }

        const value = event.target.className === "createLink" ?
                          prompt(`Insert where you would like to link to
                (make sure to include the https:// if linking to an outside site)`)
                                                              : null;

        document.execCommand(event.target.className, false, value);
    }

    /**
     * Adds a class to user-highlighted element
     *
     * @param className - name of class to add
     */
    toggleClassOnElementSelected(className) {

        const eltAnchor = window.getSelection().anchorNode;

        if (window.getSelection().anchorNode.children) { // if selected elt is an img
            window.getSelection().anchorNode.children[0].classList.toggle(className);
        }
        else {

            const eltToChange = (eltAnchor.nodeType === Node.TEXT_NODE) ? eltAnchor.parentNode : eltAnchor.lastElementChild;
            eltToChange.classList.toggle(className);
        }
    }


    render() {

        const content = React.cloneElement(this.props.children, {contentEditable: this.props.canEdit});

        if (!this.props.canEdit) {
            return <span />
        }

        return <Editable
                buttons={this.props.buttons}
                handleEdits={this.handleEdits}
                content={content}
                submit={this.props.submit}
               />
    }
}

EditableContainer.propTypes = {
    children: PropTypes.element.isRequired, //children - 1 elt to make editable
    canEdit: PropTypes.bool.isRequired,
    buttons: PropTypes.oneOf(["all", "basic", "none"]), // all buttons, or just a subset of them
    submit: PropTypes.func,
}

EditableContainer.defaultProps = {
    buttons: "all"
}


export default EditableContainer;