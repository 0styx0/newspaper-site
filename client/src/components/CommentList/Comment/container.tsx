import * as React from 'react';
import { getJWT } from '../../../helpers/jwt';
import { CommentDelete } from '../../../graphql/comment';
import Comment from './';
import { graphql, withApollo } from 'react-apollo';
import { MouseEvent } from 'react';
import graphqlErrorNotifier from '../../../helpers/graphqlErrorNotifier';

// @see Comment 's Props, which are much the same
export interface Props {
    profileLink: string;
    author: string;
    content: string;
    authorid: string;
    id: string;
    // not really void. but not using return value
    deleteComment: (params: {variables: { id: string }}) => void;
}

interface State {
    content: string;
    author: string;
}

/**
 * @param props - @see CommentContainer Props
 */
export class CommentContainer extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.onDelete = this.onDelete.bind(this);

        this.state = {
            content: this.props.content,
            author: this.props.author
        };
    }

    /**
     * Sends request to server to delete the comment
     *
     * Rerenders the Comment making it clear that it was deleted
     */
    onDelete() {

        this.setState({
            content: 'deleted',
            author: 'Deleted User'
        });

        graphqlErrorNotifier(
            this.props.deleteComment, {
                variables: {
                    id: this.props.id
                }
            },
            'commentDeleted'
        );
    }

    render() {

        const commentProps: {
            author: string;
            profileLink: string;
            content: string;
            onDelete?: (e: MouseEvent<HTMLButtonElement>) => void
        } = {
            author: this.state.author,
            profileLink: this.props.profileLink,
            content: this.state.content
        };

        const jwt = getJWT();

        // if already deleted, or user not authorized to delete,
        // or it was just created so info needed to delete isn't there yet
        if (this.props.content !== 'deleted' &&
            (jwt.level > 2 || jwt.id === this.props.authorid) &&
            this.props.id
            ) {

            commentProps.onDelete = this.onDelete;
        }

        return <Comment {...commentProps} />;
    }
}

// tslint:disable-next-line:no-any
const CommentContainerWithData = graphql(CommentDelete, {name: 'deleteComment'})(CommentContainer as any);

export default withApollo(CommentContainerWithData);