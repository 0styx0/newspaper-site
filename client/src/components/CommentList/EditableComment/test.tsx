import * as React from 'react';
import { MemoryRouter } from 'react-router';
import * as renderer from 'react-test-renderer';
import * as casual from 'casual';
import { EditableCommentContainer } from './container';
import setFakeJwt from '../../../tests/jwt.helper';
import localStorageMock from '../../../tests/localstorage.mock';
import { mount } from 'enzyme';
import * as sinon from 'sinon';

document.queryCommandSupported = () => true; // used in Editable component

describe('<EditableCommentContainer>', () => {

    const mockFunction = (() => { return; }) as any;

    function setup(id: string, functions = {addToList: mockFunction, createComment: mockFunction}) {

        return mount(
            <EditableCommentContainer
              artId={id}
              {...functions}
            />
        );
    }

    describe('snapshots', () => {

        /**
         * Takes snapshot
         */
        function snap() {

            const tree = renderer.create(
                <MemoryRouter>
                    <EditableCommentContainer
                      artId={casual.word}
                      addToList={mockFunction}
                      createComment={mockFunction}
                    />
                </MemoryRouter>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it('renders correctly when can edit', () => {

            setFakeJwt({id: casual.word});
            snap();
        });

        it('renders correctly when cannot edit', () => {

            localStorageMock.removeItem('jwt');
            snap();
        });
    });

    describe('#onSave', () => {

        beforeEach(() => setFakeJwt({id: casual.word}));

        it('saves content to state onBlur', () => {

            const wrapper = setup(casual.word);

            const content = casual.sentence;

            const contentEditable = wrapper.find('[contentEditable]').first() as any as {
                node: { innerHTML: string }, simulate: Function
            };

            contentEditable.node.innerHTML = content;
            contentEditable.simulate('blur');
            expect(wrapper.state().content).toBe(content);
        });

        it('calls commentDelete and addToList when button is clicked', () => {

            const spyComment = sinon.spy();
            const spyList = sinon.spy();

            const wrapper = setup(casual.word, {
                createComment: spyComment,
                addToList: spyList
            });

            wrapper.find('button').last().simulate('click');

            expect(spyComment.called).toBeTruthy();
            expect(spyList.called).toBeTruthy();
        });

        it('sends proper data to graphql', () => {

            const expected = {
                artId: casual.word,
                content: casual.sentence
            };

            const wrapper = setup(expected.artId, {
                createComment: (params: {variables: {artId: string, content: string}}) => {

                    expect(params.variables.artId).toBe(expected.artId);
                    expect(params.variables.content).toBe(expected.content);
                },
                addToList: mockFunction
            });

            wrapper.state().content = expected.content;
            wrapper.find('button').last().simulate('click');
        });

        it('sends content to props.addToList', () => {

            const expectedContent = casual.sentence;

            const wrapper = setup(casual.word, {
                createComment: mockFunction,
                addToList: (content: string) => expect(content).toBe(expectedContent)
            });

            wrapper.state().content = expectedContent;
            wrapper.find('button').last().simulate('click');
        });
    });

});
