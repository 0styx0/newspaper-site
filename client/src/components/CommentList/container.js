import React from 'react';
import PropTypes from 'prop-types';

import CommentContainer from './Comment/container';
import Comment from './Comment';

import EditableCommentContainer from './EditableComment/container';
import {jwt} from '../jwt';

import CommentList from './';



/**
 * @prop - array of @see Comments
 */
export default class CommentListContainer extends React.Component {

    constructor() {
        super();

        ['add'].forEach(method => this[method] = this[method].bind(this));

        this.state = {
            Comments: []
        }

    }

    /**
     * @param content - html? string
     *
     * Rerenders CommentList with EditableComment changed to Comment, and pushes a new EditableComment to the list
     */
    add(content) {

        const Comments = this.state.Comments.slice();

        Comments.push(<Comment
                        content={content}
                        authorid={jwt.id}
                        profileLink={jwt.email}
                        author="You"
                      />);

        this.setState({
            Comments
        })
    }

    // Sometimes nextProps will be undefined, which in that case, don't want to set state
    componentWillReceiveProps(nextProps) {

        if (nextProps.Comments) {

            this.setState({
                Comments: nextProps.Comments
            })
        }
    }

    shouldComponentUpdate(nextProps, nextState) {

        return !!nextProps.Comments || nextState.Comments !== this.state.Comments;
    }

    render() {

        const Comments = [...this.state.Comments];

        if (jwt.level) {
            Comments.push(<EditableCommentContainer addToList={this.add} name={this.props.name} issue={this.props.issue} />);
        }


        return <CommentList Comments={Comments}/>
    }

}




CommentListContainer.PropTypes = {

    Comments: PropTypes.arrayOf(Comment)
}
