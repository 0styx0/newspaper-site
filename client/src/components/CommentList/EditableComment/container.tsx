import * as React from 'react';
import { CommentCreate } from '../../../graphql/comment';
import EditableComment from './';
import { withApollo, graphql } from 'react-apollo';
import { FocusEvent } from 'react';
import graphqlErrorNotifier from '../../../helpers/graphqlErrorNotifier';

export interface Props {
    artId: string;
    addToList: (content: string) => void; // callback where content is passed into after user submits the comment
    createComment: (params: {variables: { content: string, artId: string }}) => void;
}

export interface State {
    content: string;
}

export class EditableCommentContainer extends React.Component<Props, State> {

    constructor(props: Props) {
        super(props);

        this.onSave = this.onSave.bind(this);

        this.state = {
            content: ''
        };
    }

    /**
     * Sends comment to be saved to database
     */
    onSave() {

        graphqlErrorNotifier(
            this.props.createComment,
            {
                variables: {
                    artId: this.props.artId,
                    content: this.state.content
                }
            },
            'commentCreated'
        )
         .then(() => this.props.addToList(this.state.content));
    }

    render() {

        return (
            <EditableComment
                onSubmit={this.onSave}
                onBlur={(e: FocusEvent<HTMLDivElement>) =>
                    this.setState({ content: (e.target as HTMLElement).innerHTML })}
            />
        );
    }

}

export default withApollo(
    // tslint:disable-next-line:no-any
    graphql(CommentCreate, {name: 'createComment'})(EditableCommentContainer as any)
);