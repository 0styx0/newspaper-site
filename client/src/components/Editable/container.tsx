import * as React from 'react';

import Editable from './';
import { MouseEvent } from 'react';

interface Props {
    children: Element | JSX.Element | HTMLElement | (Element | JSX.Element | string)[];
    canEdit: boolean;
    buttons?: 'all' | 'basic' | 'none'; // all buttons, or just a subset of them
    onSubmit?: (e: MouseEvent<HTMLButtonElement>) => void;
}

/**
 * @prop children - 1 elt to make editable
 * @prop canEdit - boolean
 * @prop buttons - if "all" (default), show all buttons. If "basic" show subset
 * @prop submit - runs when the submit button is clicked.
 *  If this is not given, nothing will happen when submit is clicked
 *
 * @return lets content be edited and renders a bar of buttons that can edit the html
 *  of props.children if props.canEdit = true
 */
export class EditableContainer extends React.Component<Props, {}> {

    constructor(props: Props) {
        super(props);

        this.handleEdits = this.handleEdits.bind(this);
    }

    handleEdits(event: MouseEvent<HTMLButtonElement>) {

        const target = event.target as HTMLElement;

        if (target.classList.contains('hideFromPreview')) {
            return this.toggleClassOnElementSelected('previewHidden');
        }

        const value = target.className === 'createLink' ?
                          prompt(`Insert where you would like to link to
                (make sure to include the https:// if linking to an outside site)`)
                                                              : null;

        document.execCommand(target.className, false, value);
    }

    /**
     * Adds a class to user-highlighted element
     *
     * @param className - name of class to add
     */
    toggleClassOnElementSelected(className: string) {

        const eltAnchor = window.getSelection().anchorNode;

        if ((window.getSelection().anchorNode as Element).children) { // if selected elt is an img
            (window.getSelection().anchorNode as Element).children[0].classList.toggle(className);

        } else {

            const eltToChange = (eltAnchor.nodeType === Node.TEXT_NODE) ?
                                  eltAnchor.parentNode as Element :
                                  (eltAnchor as Element).lastElementChild;
            eltToChange!.classList.toggle(className);
        }
    }

    render(): JSX.Element {

        const content = React.cloneElement(this.props.children as JSX.Element, {contentEditable: this.props.canEdit});

        if (!this.props.canEdit) {
            return this.props.children as JSX.Element;
        }

        return (
            <Editable
                buttons={this.props.buttons || 'all'}
                handleEdits={this.handleEdits}
                content={content}
                onSubmit={this.props.onSubmit}
            />
        );
    }
}

export default EditableContainer;
