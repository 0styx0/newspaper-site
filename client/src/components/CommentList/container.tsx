import * as React from 'react';

import { Comment } from './shared.interface';
import { getJWT } from '../jwt';

import CommentList from './';

interface Props {
    comments: Comment[];
    artId: string;
}

interface State {
    comments: Comment[];
}

/**
 * Creates a list of comments
 */
export default class CommentListContainer extends React.Component<Props, State> {

    /**
     * Seed initial state with props.comments
     */
    constructor(props: Props) {
        super(props);

        this.onAdd = this.onAdd.bind(this);

        this.state = {
            comments: props.comments
        };

    }

    /**
     * @param content - html? string
     *
     * Rerenders CommentList with EditableComment changed to Comment, and pushes a new EditableComment to the list
     */
    onAdd(content: string) {

        const comments = this.state.comments.slice();
        const jwt = getJWT();

        comments.push({
                content,
                author: {
                    profileLink: jwt.profileLink,
                    fullName: 'You',
                    id: jwt.id
                },
                canDelete: true,
                dateCreated: (new Date()).toISOString(),
                id: new Date().toString() /* random temp id*/
        });

        this.setState({
            comments
        });
    }

    render() {

        return (
            <CommentList
              comments={this.state.comments}
              canAdd={!!getJWT().level}
              onAdd={this.onAdd}
              artId={this.props.artId}
            />
        );
    }
}
