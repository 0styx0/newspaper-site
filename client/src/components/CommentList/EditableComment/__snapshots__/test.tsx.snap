// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableCommentContainer> snapshots renders correctly when can edit 1`] = `
<article
  className="comment"
  id="reply"
>
  <div>
    <div
      id=""
    >
      <button
        className="bold"
        onClick={[Function]}
        type="button"
      >
        bold
      </button>
      <button
        className="italic"
        onClick={[Function]}
        type="button"
      >
        italic
      </button>
      <button
        className="underline"
        onClick={[Function]}
        type="button"
      >
        underline
      </button>
      <button
        className="strikeThrough"
        onClick={[Function]}
        type="button"
      >
        strikeThrough
      </button>
      <button
        className="createLink"
        onClick={[Function]}
        type="button"
      >
        createLink
      </button>
      <br />
      <button
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </div>
    <div
      className="content"
      contentEditable={true}
      onBlur={[Function]}
    />
  </div>
</article>
`;

exports[`<EditableCommentContainer> snapshots renders correctly when cannot edit 1`] = `
<article
  className="comment"
  id="reply"
>
  <div
    className="content"
    onBlur={[Function]}
  />
</article>
`;
